bin/desi_use_reservation

```diff
@@ -0,0 +1,211 @@
+    jobs = get_jobs_in_queue()
+
+    #- Sort by name so that earlier nights are prioritized over later nights
+    jobs.sort('NAME')
+
+    #- Filter which jobs are in reservation vs. eligible to move into reservation
+    jobs_in_reservation = jobs[jobs['RESERVATION'] == name]
+
+    eligible_for_reservation = jobs['RESERVATION'] == 'null'
+    eligible_for_reservation &= jobs['PARTITION'] == regular_partition
+
+    #- Only move jobs that are currently eligible to run, not those waiting for dependencies
+    #- so that we don't fill reservation with jobs that can't run
+    eligible_for_reservation &= jobs['ST'] == 'PD'
+    eligible_for_reservation &= jobs['NODELISTREASON'] != 'Dependency'
+    jobs_eligible = jobs[eligible_for_reservation]
+
+    #- Counting jobs and nodes in and out of the reservation
+    njobs_in_reservation = len(jobs_in_reservation)
+    njobnodes_in_reservation = np.sum(jobs_in_reservation['NODES'])
+
+    njobs_eligible = len(jobs_eligible)
+    njobnodes_eligible = np.sum(jobs_eligible['NODES'])
+
+    if njobs_eligible == 0:
+        log.info('No available jobs to add')
+        return
+
+    njobnodes_to_add = max(0, ressize + extra_nodes - njobnodes_in_reservation)
+    njobnodes_to_add = min(njobnodes_to_add, njobnodes_eligible)
+
+    log.info(f'At {time.asctime()}, {name} ({ressize} nodes) has {njobs_in_reservation} jobs using {njobnodes_in_reservation} nodes')
+    log.info(f'{njobs_eligible} {restype} jobs using {njobnodes_eligible} nodes are eligible to be moved into the reservation')
+
+    if njobnodes_to_add == 0:
+        log.info('Reservation full, no need to add more jobs at this time')
+        return
+
+    log.info(f'Adding jobs to use {njobnodes_to_add} additional nodes')
+    jobnodes_added = 0
+    jobids = list()
+    for jobtype in jobtypes:
+        ii = np.char.startswith(jobs_eligible['NAME'], jobtype)
+        for row in jobs_eligible[ii]:
+            jobname = row['NAME']
+            log.info(f'Move {jobname} to {name}')
+            jobnodes_added += row['NODES']
+            jobids.append(row['JOBID'])
+
+            if jobnodes_added >= njobnodes_to_add:
+                break
+
+        if jobnodes_added >= njobnodes_to_add:
+            break
+
+    if len(jobids) > 0:
+        if dry_run:
+            log.info('Dry run mode; will print what to do but not actually run the commands')
+        else:
+            log.info('Running scontrol commands')
+
+        #- Update queue in batches of batch_size jobs
+        batch_size = 10
+        for i in range(0, len(jobids), batch_size):
+            jobids_csv = ','.join([str(x) for x in jobids[i:i+batch_size]])
+            cmd = f'scontrol update ReservationName={name} JobID={jobids_csv}'
+            if dry_run:
+                #- Purposefully print, not log, to make it easier to cut-and-paste
+                print(cmd)
+            else:
+                log.info(cmd)
+                subprocess.run(cmd.split())
```

Comment:
Since you are determining the jobs earlier and then executing them here there is a small chance that a job will start running before the `subprocess` `squeue` call can execute. I think that will raise an error. I would wrap this in a `try`-`except` loop to keep it from crashing. 

Granted that this should take a tiny fraction of a second to compute before execution, but with large numbers rare things happen.

This is especially important since one of the modes here is a long-running daemon-like job.

Reply:
Short version: done.

Long version:
I explicitly checked what would happen in this case.  By default, `subprocess.run(cmd.split())` does *not* raise an exception, but rather returns an object with info about whether the command succeeded or failed.  I changed this to `subprocess.run(cmd.split(), check=True)` which causes it to raise a `subprocess.CalledProcessException` upon failure.  I catch that and log the error, but don't take other action on this loop, just leaving it for a new loop with new jobs to likely ignore it (this case) or try again (other generic slurm hiccups).